File: [0907][Medium][Sum_of_Subarray_Minimums][Stack_Array]/Sum_of_Subarray_Minimums.cpp
// Time  Complexity: O(N) 
// Space Complexity: O(N)
// solution 1, 单调栈 Monotone Stack,construct preMinArr[],nextMinArr[]
#include <stack>
#include <vector>
using namespace std;
class Solution {
public:
  int sumSubarrayMins(vector<int>& A) {
    stack<pair<int, int>> in_stk_prev, in_stk_next;
    // left is for the distance to previous less element
    // right is for the distance to next less element
    vector<int> left(A.size()), right(A.size());
		
    //initialize
    for(int i = 0; i < A.size(); i++) left[i] =  i + 1;
    for(int i = 0; i < A.size(); i++) right[i] = A.size() - i;
		
    for(int i = 0; i < A.size(); i++){
      // for previous less
      while(!in_stk_prev.empty() && in_stk_prev.top().first > A[i]) 
        in_stk_prev.pop();
      left[i] = in_stk_prev.empty()? i + 1: i - in_stk_prev.top().second;
      in_stk_prev.push({A[i],i});
			
      // for next less
      while(!in_stk_next.empty() && in_stk_next.top().first > A[i]){
        auto x = in_stk_next.top();
        in_stk_next.pop();
        right[x.second] = i - x.second;
      }
      in_stk_next.push({A[i], i});
    }

    int ans = 0, mod = 1e9 +7;
    for(int i = 0; i < A.size(); i++){
      ans = (ans + A[i]*left[i]*right[i])%mod;
    }
    return ans;
  }
};

File: [0907][Medium][Sum_of_Subarray_Minimums][Stack_Array]/Sum_of_Subarray_Minimums.py
# Time  Complexity: O(N) 
# Space Complexity: O(N)
# solution 1, 单调栈 Monotone Stack, construct preMinArr[],nextMinArr[]
class Solution(object):
    def sumSubarrayMins(self, A):
        MOD = 10**9 + 7
        N = len(A)

        # prev has i* - 1 in increasing order of A[i* - 1]
        # where i* is the answer to query j
        stack = []
        prev = [None] * N
        for i in range(N):
            while stack and A[i] <= A[stack[-1]]:
                stack.pop()
            prev[i] = stack[-1] if stack else -1
            stack.append(i)

        # next has k* + 1 in increasing order of A[k* + 1]
        # where k* is the answer to query j
        stack = []
        next_ = [None] * N
        for k in range(N-1, -1, -1):
            while stack and A[k] < A[stack[-1]]:
                stack.pop()
            next_[k] = stack[-1] if stack else N
            stack.append(k)

        # Use prev/next array to count answer
        return sum((i - prev[i]) * (next_[i] - i) * A[i]
                   for i in range(N)) % MOD

def main():
    nums = [3,1,2,4]      #expect is 17
    obj = Solution()
    res = obj.sumSubarrayMins(nums)
    print("return value is ",res);
    
if __name__ =='__main__':
    main()   